File: go/poller.go
package mast

import (
	"context"
	"errors"
	"sort"
	"time"
)

// Age of an open page against how often to ask about it. The first rung the
// page has not aged past wins: fast while somebody may still be looking at it,
// slow once it is clear nobody is.
var pollLadder = []struct {
	Until time.Duration
	Every time.Duration
}{
	{30 * time.Second, 3 * time.Second},
	{5 * time.Minute, 10 * time.Second},
	{0, 30 * time.Second},
}

const pollTick = time.Second

// statuses is all a Poller needs of a Channel, which is also what makes the
// ladder and the budget testable without a socket.
type statuses interface {
	Status(ctx context.Context, messageID string) (Status, error)
}

// Poller watches open pages on one channel until they settle.
//
// Several at once share the channel's request budget, so the oldest is asked
// about first when there is not enough to go round: it is the one somebody is
// still standing over. A 429 backs off the whole channel rather than the one
// page, because the limit belongs to the key.
type Poller struct {
	// Budget is how many requests a minute polling may spend. The rest of the
	// channel's allowance is left for sending.
	Budget int

	channel statuses
	open    map[string]*watched
	last    map[string]Status
	spent   []time.Time
	paused  time.Time

	now   func() time.Time
	sleep func(context.Context, time.Duration) error
}

type watched struct {
	messageID string
	started   time.Time
	lastPoll  time.Time
}

func (w *watched) due(at time.Time) bool {
	age := at.Sub(w.started)
	for _, rung := range pollLadder {
		if rung.Until == 0 || age < rung.Until {
			return at.Sub(w.lastPoll) >= rung.Every
		}
	}
	return false
}

// NewPoller watches pages on one channel, spending half the channel's request
// allowance.
func NewPoller(c *Channel) *Poller {
	return &Poller{
		Budget:  PollBudgetPerMinute,
		channel: c,
		open:    map[string]*watched{},
		last:    map[string]Status{},
		now:     time.Now,
		sleep:   sleep,
	}
}

// Watch starts following a message. Watching one twice watches one page.
func (p *Poller) Watch(messageID string) error {
	if !messageIDPattern.MatchString(messageID) {
		return &Rejected{Reason: messageID + " is not a message id"}
	}
	if _, ok := p.open[messageID]; !ok {
		p.open[messageID] = &watched{messageID: messageID, started: p.now()}
	}
	return nil
}

// Open is how many pages are still being watched.
func (p *Poller) Open() int { return len(p.open) }

// Last is the most recent status read for a message.
func (p *Poller) Last(messageID string) (Status, bool) {
	status, ok := p.last[messageID]
	return status, ok
}

// Tick asks about whatever is due and affordable, and returns what settled.
func (p *Poller) Tick(ctx context.Context) (map[string]Status, error) {
	settled := map[string]Status{}
	at := p.now()
	if at.Before(p.paused) {
		return settled, nil
	}

	var due []*watched
	for _, page := range p.open {
		if page.due(at) {
			due = append(due, page)
		}
	}
	sort.Slice(due, func(i, j int) bool { return due[i].started.Before(due[j].started) })

	for _, page := range due {
		if !p.take(at) {
			break
		}
		page.lastPoll = at
		status, err := p.channel.Status(ctx, page.messageID)
		if err != nil {
			var limited *RateLimited
			switch {
			case errors.As(err, &limited):
				p.paused = p.now().Add(limited.RetryAfter)
				return settled, nil
			case errors.Is(err, ErrUnknownMessage), errors.Is(err, ErrUnknownChannel):
				// Nothing more is going to happen to a message the channel
				// will not talk about.
				delete(p.open, page.messageID)
			case ctx.Err() != nil:
				return settled, ctx.Err()
			default:
				// A blip is not an answer: leave the page open for the next
				// tick.
			}
			continue
		}

		p.last[page.messageID] = status
		if status.Done() {
			delete(p.open, page.messageID)
			settled[page.messageID] = status
		}
	}
	return settled, nil
}

// Run ticks until nothing is open or ctx is done. Whatever settled first is
// returned alongside ctx's error, and anything still open stays watched, so
// calling Run again picks up where this left off. A deadline here is not a
// failed page: it is still open on the phone and can still be answered.
func (p *Poller) Run(ctx context.Context) (map[string]Status, error) {
	settled := map[string]Status{}
	for len(p.open) > 0 {
		done, err := p.Tick(ctx)
		for id, status := range done {
			settled[id] = status
		}
		if err != nil {
			return settled, err
		}
		if len(p.open) == 0 {
			break
		}
		if err := p.sleep(ctx, pollTick); err != nil {
			return settled, err
		}
	}
	return settled, nil
}

// take spends one request if the last sixty seconds leave room for it.
//
// A token bucket would be the usual answer, but a full one lets through its
// own depth on top of the refill, which is how a limiter sized at half the
// channel's allowance ends up spending nearly all of it.
func (p *Poller) take(at time.Time) bool {
	cutoff := at.Add(-time.Minute)
	kept := p.spent[:0]
	for _, when := range p.spent {
		if when.After(cutoff) {
			kept = append(kept, when)
		}
	}
	p.spent = kept
	if len(p.spent) >= p.Budget {
		return false
	}
	p.spent = append(p.spent, at)
	return true
}

func sleep(ctx context.Context, d time.Duration) error {
	timer := time.NewTimer(d)
	defer timer.Stop()
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-timer.C:
		return nil
	}
}

File: rust/src/poller.rs
use std::collections::{HashMap, VecDeque};
use std::time::{Duration, Instant};

use crate::transport::BoxFuture;
use crate::{Channel, Error, Status, POLL_BUDGET_PER_MINUTE};

/// Age of an open page against how often to ask about it. The first rung the
/// page has not aged past wins: fast while somebody may still be looking at it,
/// slow once it is clear nobody is.
const POLL_LADDER: [(f64, f64); 3] = [(30.0, 3.0), (300.0, 10.0), (f64::INFINITY, 30.0)];

/// How long [`run`](Poller::run) waits between looks at the ladder.
#[cfg(feature = "tokio")]
const TICK: Duration = Duration::from_secs(1);

type Clock = Box<dyn Fn() -> f64 + Send + Sync>;
type Sleeper = Box<dyn Fn(Duration) -> BoxFuture<'static, ()> + Send + Sync>;

struct Watched {
    started: f64,
    last_poll: f64,
}

impl Watched {
    fn due(&self, at: f64) -> bool {
        let age = at - self.started;
        for (until, every) in POLL_LADDER {
            if age < until {
                return at - self.last_poll >= every;
            }
        }
        false
    }
}

/// Watches open pages on one channel until they settle.
///
/// Several at once share the channel's request budget, so the oldest is asked
/// about first when there is not enough to go round: it is the one somebody is
/// still standing over. A 429 backs off the whole channel rather than the one
/// page, because the limit belongs to the key.
pub struct Poller<'a> {
    channel: &'a Channel,
    budget: usize,
    open: HashMap<String, Watched>,
    last: HashMap<String, Status>,
    dropped: HashMap<String, Error>,
    spent: VecDeque<f64>,
    paused_until: f64,
    clock: Clock,
    sleeper: Option<Sleeper>,
}

impl<'a> Poller<'a> {
    pub fn new(channel: &'a Channel) -> Poller<'a> {
        let started = Instant::now();
        Poller {
            channel,
            budget: POLL_BUDGET_PER_MINUTE,
            open: HashMap::new(),
            last: HashMap::new(),
            dropped: HashMap::new(),
            spent: VecDeque::new(),
            paused_until: 0.0,
            clock: Box::new(move || started.elapsed().as_secs_f64()),
            sleeper: None,
        }
    }

    /// How many requests a minute polling may spend. The rest of the channel's
    /// allowance is left for sending.
    pub fn with_budget(mut self, per_minute: usize) -> Self {
        self.budget = per_minute;
        self
    }

    /// Monotonic seconds from any source, for a caller that already keeps time
    /// or a test that would rather not wait.
    pub fn with_clock(mut self, clock: impl Fn() -> f64 + Send + Sync + 'static) -> Self {
        self.clock = Box::new(clock);
        self
    }

    /// What [`run`](Poller::run) waits with between ticks.
    pub fn with_sleeper(
        mut self,
        sleeper: impl Fn(Duration) -> BoxFuture<'static, ()> + Send + Sync + 'static,
    ) -> Self {
        self.sleeper = Some(Box::new(sleeper));
        self
    }

    /// Start following a message. Watching one twice watches one page.
    pub fn watch(&mut self, message_id: &str) -> Result<(), Error> {
        if !crate::url::is_message_id(message_id) {
            return Err(Error::Rejected(format!(
                "{message_id:?} is not a message id"
            )));
        }
        let at = (self.clock)();
        self.open.entry(message_id.to_string()).or_insert(Watched {
            started: at,
            last_poll: f64::NEG_INFINITY,
        });
        Ok(())
    }

    pub fn open_count(&self) -> usize {
        self.open.len()
    }

    /// The most recent status read for a message.
    pub fn last(&self, message_id: &str) -> Option<&Status> {
        self.last.get(message_id)
    }

    /// Why a message stopped being watched without settling. Taking it clears
    /// it, so a caller that watches the same id again starts clean.
    pub fn take_error(&mut self, message_id: &str) -> Option<Error> {
        self.dropped.remove(message_id)
    }

    /// Ask about whatever is due and affordable, and report what settled.
    pub async fn tick(&mut self) -> Result<HashMap<String, Status>, Error> {
        let mut settled = HashMap::new();
        let at = (self.clock)();
        if at < self.paused_until {
            return Ok(settled);
        }

        let mut due: Vec<(String, f64)> = self
            .open
            .iter()
            .filter(|(_, page)| page.due(at))
            .map(|(id, page)| (id.clone(), page.started))
            .collect();
        due.sort_by(|a, b| a.1.total_cmp(&b.1));

        for (message_id, _) in due {
            if !self.take(at) {
                break;
            }
            if let Some(page) = self.open.get_mut(&message_id) {
                page.last_poll = at;
            }

            match self.channel.status(&message_id).await {
                Ok(status) => {
                    let done = status.done();
                    self.last.insert(message_id.clone(), status.clone());
                    if done {
                        self.open.remove(&message_id);
                        settled.insert(message_id, status);
                    }
                }
                Err(Error::RateLimited { retry_after }) => {
                    self.paused_until = (self.clock)() + retry_after.as_secs_f64();
                    break;
                }
                Err(err @ (Error::UnknownMessage(_) | Error::UnknownChannel(_))) => {
                    // Nothing more is going to happen to a message the channel
                    // will not talk about.
                    self.open.remove(&message_id);
                    self.dropped.insert(message_id, err);
                }
                Err(_) => {
                    // A blip is not an answer: leave the page open for the next
                    // tick.
                }
            }
        }
        Ok(settled)
    }

    /// Tick until nothing is open or the timeout runs out. A timeout is not a
    /// failed page: anything still open stays watched, so calling this again
    /// picks up where it left off.
    #[cfg(feature = "tokio")]
    pub async fn run(
        &mut self,
        timeout: Option<Duration>,
    ) -> Result<HashMap<String, Status>, Error> {
        let deadline = timeout.map(|limit| (self.clock)() + limit.as_secs_f64());
        let mut settled = HashMap::new();

        while !self.open.is_empty() {
            settled.extend(self.tick().await?);
            if self.open.is_empty() {
                break;
            }
            if deadline.is_some_and(|deadline| (self.clock)() >= deadline) {
                break;
            }
            match &self.sleeper {
                Some(sleeper) => sleeper(TICK).await,
                None => tokio::time::sleep(TICK).await,
            }
        }
        Ok(settled)
    }

    /// Spend one request if the last sixty seconds leave room for it.
    ///
    /// A token bucket would be the usual answer, but a full one lets through
    /// its own depth on top of the refill, which is how a limiter sized at half
    /// the channel's allowance ends up spending nearly all of it.
    fn take(&mut self, at: f64) -> bool {
        while self.spent.front().is_some_and(|spent| at - spent >= 60.0) {
            self.spent.pop_front();
        }
        if self.spent.len() >= self.budget {
            return false;
        }
        self.spent.push_back(at);
        true
    }
}
